sdk/core/core-https/src/policies/bearerTokenAuthenticationPolicy.ts

```diff
@@ -34,26 +73,73 @@ export function bearerTokenAuthenticationPolicy(
 ): PipelinePolicy {
   const { credential, scopes } = options;
   const tokenCache: AccessTokenCache = new ExpiringAccessTokenCache();
-  async function getToken(tokenOptions: GetTokenOptions): Promise<string | undefined> {
+  const { prepareRequest, getChallenge = defaultGetChallenge, processChallenge } =
+    options.challenge || {};
+
+  /**
+   * getToken will call to the underlying credential's getToken request with properties coming from the request,
+   * as well as properties coming from parsing the CAE challenge, if any.
+   */
+  async function getToken(
+    request: PipelineRequest,
+    context?: AuthenticationContext
+  ): Promise<string | undefined> {
     let accessToken = tokenCache.getCachedToken();
+    const getTokenOptions: GetTokenOptions = {
+      claims: context?.claims,
+      abortSignal: request.abortSignal,
+      tracingOptions: {
+        spanOptions: request.spanOptions
+      }
+    };
     if (accessToken === undefined) {
-      accessToken = (await credential.getToken(scopes, tokenOptions)) || undefined;
+      accessToken =
+        (await credential.getToken(context?.scopes || scopes, getTokenOptions)) || undefined;
       tokenCache.setCachedToken(accessToken);
     }
-
     return accessToken ? accessToken.token : undefined;
   }
+
+  /**
+   * Populates the token in the request headers.
+   */
+  function assignToken(request: PipelineRequest, token?: string): PipelineRequest {
+    request.headers.set("Authorization", `Bearer ${token}`);
+    return request;
+  }
+
+  /**
+   * Uses the challenge parameters to:
+   * - Prepare the outgoing request (if the `prepareRequest` method has been provided).
+   * - Process a challenge if the response contains it.
+   * - Retrieve a token with the challenge information, then re-send the request.
+   */
+  async function challengePolicy(
+    request: PipelineRequest,
+    next: SendRequest
+  ): Promise<PipelineResponse> {
+    if (prepareRequest) {
+      await prepareRequest(request);
+    }
+    const response = await next(request);
+    const challenge = getChallenge(response);
+    if (challenge && processChallenge) {
+      const context = await processChallenge(challenge);
+      const token = await getToken(request, context);
+      return next(assignToken(request, token));
+    }
+    return response;
+  }
+
   return {
     name: bearerTokenAuthenticationPolicyName,
     async sendRequest(request: PipelineRequest, next: SendRequest): Promise<PipelineResponse> {
-      const token = await getToken({
-        abortSignal: request.abortSignal,
-        tracingOptions: {
-          spanOptions: request.spanOptions
-        }
-      });
-      request.headers.set("Authorization", `Bearer ${token}`);
-      return next(request);
+      const token = await getToken(request);
+      if (token) {
+        return next(assignToken(request, token));
```

Comment:
I have a question here - what happens if the server issues a challenge even though we're providing a token? Or does that never happen?

Reply:
That isn't part of the protocol! But I have changed this to the following:

```ts
      let accessToken = tokenCache.getCachedToken();
      if (!accessToken && processChallenge) {
        return await challengePolicy(request, next);
      } else {
        const token = await retrieveToken(request);
        return next(assignToken(request, token));
      }
```

Since it only makes sense to process the challenge if there's no cached token 🌞 how does that sound?

Reply:
We should process the challenge anytime we receive it! Nice catch! I've fixed this here: https://github.com/Azure/azure-sdk-for-js/pull/13888/commits/435776e50b77acd2dc2881dfa9761766562364d1